### PR TITLE
Add the option to build with docker

### PR DESCRIPTION
**Pull request by Anon7250**
This adds a new way to build `libgocryptfs` without having to set up Android NDK or Golang. The hope is that this simplifies the setup of a development environment for a new developer.

### CODE REVIEW

**Comment by hardcore-sushi** [comment]
Good idea, but I have some questions:

- Is it really needed to start the build process as root ?
- Wouldn't it be better if we used a shared volume instead of copying the whole repository ? That way, we don't need to rebuild OpenSSL each time. We could also build libgocryptfs for only some architectures and directly get the result, instead of making an archive, copying then extracting it.
- Wouldn't it be more secure if we let the user download the OpenSSL sources ? It would allow him to pick the latest version and verify OpenPGP signature.

**Comment by Anon7250** [comment]
Yup, these are all good questions. I'll modify the scripts.

**Comment by Anon7250** [comment]
It took me a while (sorry). I've attempted to address your questions. Please let me know how this looks. Thank you!

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,11 @@
+FROM smithmicro/android-build:28-ndk21
+
+USER root
+RUN apt-get update \
+    && apt-get install --no-install-recommends -y golang make gcc libcrypto++-dev libssl-dev pkg-config \
+    && groupadd builder && useradd -g builder builder \
+    && mkdir -p /home/builder && chown -R builder:builder /home/builder
```
Comment (hardcore-sushi) [inline Dockerfile]: I think this 2 lines could be replaced by:
```
useradd -m builder
```

File: build.sh [hunk 2]
```diff
@@ -1,5 +1,8 @@
 #!/bin/bash
 
+if [ ${FORCE_INSTALL_GO_DEPS} -eq 1 ]; then
+  go get golang.org/x/sys/unix golang.org/x/sys/cpu golang.org/x/crypto/hkdf github.com/jacobsa/crypto/siv github.com/rfjakob/eme
```
Comment (hardcore-sushi) [inline build.sh]: Why installing the go dependencies at run time rather than when building the container?

File: build-with-docker.sh [hunk 3]
```diff
@@ -0,0 +1,8 @@
+set -e
+if [ ! -f openssl.tar ]; then
```
Comment (hardcore-sushi) [inline build-with-docker.sh]: Sorry but I don't understand this line.

**Comment by Anon7250** [comment]
Oops I only saw the last comment about `openssl.tar`. I'll address the other two issues as well.

**Comment by Anon7250** [comment]
Thanks for the suggestions!

File: build-with-docker.sh [hunk 4]
```diff
@@ -0,0 +1,8 @@
+set -e
+if [ ! -d openssl ]; then
+    echo "Before we continue, please untar OpenSSL source code to libgocryptfs/openssl/"
+    exit 1
+fi
+docker build . -t libgocryptfs
+docker container rm libgocryptfs-instance
```
Comment (hardcore-sushi) [inline build-with-docker.sh:7]: This line won't work the first time we run the script. Maybe it should be better to remove it and run the container with `--rm`.

File: build-with-docker.sh [hunk 5]
```diff
@@ -0,0 +1,8 @@
+set -e
+if [ ! -d openssl ]; then
```
Comment (hardcore-sushi) [inline build-with-docker.sh:2]: Maybe we could replace the hardcoded path with a value passed in argument so we don't need to rename the openssl folder.

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,11 @@
+FROM smithmicro/android-build:28-ndk21
+
+USER root
+RUN apt-get update \
+    && apt-get install --no-install-recommends -y golang make gcc libcrypto++-dev libssl-dev pkg-config \
+    && groupadd builder && useradd -m -g builder builder
```
Comment (hardcore-sushi) [inline Dockerfile:6]: We don't need to create the group manually here, `useradd` will do it automatically.

**Comment by hardcore-sushi** [comment]
I tested your scripts but I had problems with go. I think previous versions don't like projects without a `go.mod` file. I will fix it first then merge your PR.
Also, I think it would be smart to build the container from an empty directory so that docker doesn't add the whole libgocryptfs root folder to the build context.

**Comment by Anon7250** [comment]
> I tested your scripts but I had problems with go.

That's surprising. I was able to generate `libgocryptfs.so` by running `build-with-docker.bat`. What's the error message? Did it show up during `docker build` or `docker run`?

> I think it would be smart to build the container from an empty directory so that docker doesn't add the whole libgocryptfs root folder to the build context.

This is do-able. Just to double check, we should mount `openssl` directory so that we don't recompile openssl every time, and for the rest of the source code, we should create a copy instead of mounting them. What about the resulting binary files? Should we generate them in a mount point, or copy them out?

**Comment by hardcore-sushi** [comment]
> What's the error message? Did it show up during `docker build` or `docker run`?

During execution of `build.sh`:
```
go: go.mod file not found in current directory or any parent directory; see 'go help modules'
```
I used go 1.16.5 in the container. I don't have this error with go 1.16.9 on my host. I will create a `go.mod` file so every go versions will be happy.
> I think it would be smart to build the container from an empty directory so that docker doesn't add the whole libgocryptfs root folder to the build context.

I mean, when doing `docker build -t libgocryptfs .`, we don't need any files so we could run this from an empty directory to prevent this: `Sending build context to Docker daemon  153.1MB`. Once the container is built, we could start it from the root directory to mount the source code.

**Comment by Anon7250** [comment]
> go: go.mod file not found in current directory or any parent directory; see 'go help modules'

The container built by my docker instance uses go 1.10.4. I was expecting that docker will install the same version of go for everyone, so they don't have to debug issues with `go.mod`.

I couldn't reproduce this error, but here is a screenshot of my `go version` output, and a video screencap of the entire build process.

go version


https://user-images.githubusercontent.com/74374523/138572529-b4fbc9c2-21d3-488d-927b-e01db86917ac.mp4

(The compression on this video is horrible due to the file size limit)

-----

My next step is to address your other comments. Let me know how you would like to proceed in terms of go errors.

**Comment by hardcore-sushi** [comment]
Yeah I used `golang:1.16-alpine` because I cannot download the entire `smithmicro/android-build:28-ndk21` image. But now it's fixed: [1973153](https://github.com/hardcore-sushi/libgocryptfs/commit/19731536027d0b64c19742846bf8dda6920110ba).

Also, you should no longer need to run the `go get` command, dependencies should automatically be downloaded at compile time.

**Pull request closed without merging**